packages/user_packages/iotcc_mqtt.py

```diff
@@ -76,50 +99,73 @@ def run(self, registry):
-                    if attempts == max_retry_attempts:
+                except Exception, e:
+                    if reg_attempts == retry_attempts:
                         raise
-                    attempts += 1
-                    # The sleep time before re-trying depends on the infrastructure requirement of broker to restart
-                    # It can be modified or removed as per the infrastructure requirement
-                    time.sleep(5)
+                    reg_attempts += 1
+                    log.error('MQTT Connection Failed - {0}'.format(str(e)))
+                    log.info('Trying Edge System {0} Registration: Attempt - {1}'.format(edge_system.name,
+                                                                                         str(reg_attempts)))
+                    time.sleep(delay_retries)
+
             registry.register("iotcc_mqtt", self.iotcc)
-            # Store the registered edge system object in liota package manager registry after the
             registry.register("iotcc_mqtt_edge_system", self.iotcc_edge_system)
 
-        except RegistrationFailure:
-            print "EdgeSystem registration to IOTCC failed"
+            # Attempts for setting edge system properties
+            prop_attempts = 0
+            # Set Properties with retry attempts in case of exception
+            while prop_attempts < retry_attempts:
+                try:
+                    self.iotcc.set_properties(self.iotcc_edge_system,
+                                              {"Country": "USA-G", "State": "California", "City": "Palo Alto",
+                                               "Location": "VMware HQ", "Building": "Promontory H Lab",
+                                               "Floor": "First Floor"})
+                    break
+                except Exception, e:
+                    prop_attempts += 1
+                    log.error(
+                        'Exception while setting Property for Edge System {0} - {1}'.format(edge_system.name, str(e)))
+                    log.info('Trying setting properties for Edge System {0}: Attempt - {1}'.format(edge_system.name,
+                                                                                                   str(prop_attempts)))
+                    time.sleep(delay_retries)
+
+        except Exception:
+            log.error("EdgeSystem registration to IOTCC failed even after all the retries starting cleanup process")
+            self.clean_up()
+            raise
 
     def clean_up(self):
-        """
-        The clean up function of a liota package.
-
-        Disconnects from IoTControlCenter DCC and un-registers the edge-system.
-
-        :return:
-        """
-        # Unregister the edge system on package unload
-        # Kindly include the edge system un-register call on package unload
+        # Unregister the edge system
         self.iotcc.unregister(self.iotcc_edge_system)
```

Comment:
If somebody wants to unload the package for time being and has unregister call, it will end up deleting the history of the edge system/devices from Pulse IoT Center. Please add a comment as to when it is needed to have unregister() call and when it is not.

Reply:
Added